internal/libyaml/yamldatatest_loader.go

```diff
@@ -0,0 +1,503 @@
+			// For now, skip aliases and comments (not used in test data)
+		}
+	}
+
+	return root, nil
+}
+
+// UnmarshalStruct populates a struct from a map using reflection and yaml tags.
+// This function is exported so it can be used by other packages for data-driven testing.
+//
+// Example:
+//
+//	type TestCase struct {
+//	    Name string `yaml:"name"`
+//	    Data string `yaml:"data"`
+//	}
+//	var tc TestCase
+//	err := libyaml.UnmarshalStruct(&tc, map[string]interface{}{
+//	    "name": "test1",
+//	    "data": "hello",
+//	})
+func UnmarshalStruct(target interface{}, data map[string]interface{}) error {
+	v := reflect.ValueOf(target)
+	if v.Kind() != reflect.Ptr || v.Elem().Kind() != reflect.Struct {
+		return fmt.Errorf("target must be pointer to struct, got %T", target)
+	}
+
+	v = v.Elem()
+	t := v.Type()
+
+	// Build map of yaml tag names to field indices (support multiple fields per tag)
+	fieldMap := make(map[string][]int)
+	for i := 0; i < t.NumField(); i++ {
+		field := t.Field(i)
+		yamlTag := field.Tag.Get("yaml")
+		if yamlTag != "" && yamlTag != "-" {
+			// Remove options like ",omitempty"
+			if idx := strings.Index(yamlTag, ","); idx != -1 {
+				yamlTag = yamlTag[:idx]
+			}
+			fieldMap[yamlTag] = append(fieldMap[yamlTag], i)
+		}
+	}
+
+	// Populate fields from map
+	for key, value := range data {
+		fieldIndices, ok := fieldMap[key]
+		if !ok {
+			// Skip unknown fields
+			continue
+		}
+
+		// Try each field with this tag until one succeeds
+		var lastErr error
+		for _, fieldIdx := range fieldIndices {
+			field := v.Field(fieldIdx)
+			if !field.CanSet() {
+				continue
+			}
+
+			if err := setField(field, value); err != nil {
+				lastErr = err
+				continue
+			}
+			// Success, move to next key
+			lastErr = nil
+			break
+		}
+
+		// If all fields failed, return the last error
+		if lastErr != nil {
+			return fmt.Errorf("field %s: %w", key, lastErr)
+		}
+	}
+
+	return nil
+}
+
+// setField sets a reflect.Value from an interface{} value
+func setField(field reflect.Value, value interface{}) error {
+	if value == nil {
+		// Set zero value
+		field.Set(reflect.Zero(field.Type()))
+		return nil
+	}
+
+	fieldType := field.Type()
+
+	// Handle custom types with FromValue method (IntOrStr, ByteInput, Args)
+	if field.CanAddr() {
+		if converter, ok := field.Addr().Interface().(interface{ FromValue(interface{}) error }); ok {
+			if err := converter.FromValue(value); err != nil {
+				return err
+			}
+			return nil
+		}
+	}
+
+	// Handle basic types
+	valueRefl := reflect.ValueOf(value)
+	valueType := valueRefl.Type()
+
+	// Direct assignment if types match
+	if valueType.AssignableTo(fieldType) {
+		field.Set(valueRefl)
+		return nil
+	}
+
+	// Handle conversions
+	switch fieldType.Kind() {
+	case reflect.String:
+		if str, ok := value.(string); ok {
+			field.SetString(str)
+			return nil
+		}
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		if intVal, ok := value.(int); ok {
```

Comment:
Int64 will overflow int on i386 arch here (int32)

CanInt and/or OverflowInt should be considered